minicluster/src/main/java/alluxio/external/ExternalCluster.java

```diff
@@ -0,0 +1,403 @@
+import alluxio.AlluxioURI;
+import alluxio.Configuration;
+import alluxio.ConfigurationRule;
+import alluxio.ConfigurationTestUtils;
+import alluxio.PropertyKey;
+import alluxio.cli.Format;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.FileSystemContext;
+import alluxio.exception.status.UnavailableException;
+import alluxio.master.MasterInquireClient;
+import alluxio.master.SingleMasterInquireClient;
+import alluxio.master.ZkMasterInquireClient;
+import alluxio.network.PortUtils;
+import alluxio.util.CommonUtils;
+import alluxio.util.network.NetworkAddressUtils;
+
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+import org.apache.commons.io.Charsets;
+import org.apache.commons.io.FileUtils;
+import org.apache.curator.test.TestingServer;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ThreadLocalRandom;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Class for starting, stopping, and interacting with an Alluxio cluster where each master and
+ * worker runs in its own process.
+ *
+ * The synchronization strategy for this class is to synchronize all public methods.
+ */
+@ThreadSafe
+public final class ExternalCluster implements TestRule {
+  private static final Logger LOG = LoggerFactory.getLogger(ExternalCluster.class);
+  private static final File ARTIFACTS_DIR = new File("./target/artifacts");
+  private static final File TESTS_LOG = new File("./target/tests.log");
+
+  private final Map<PropertyKey, String> mProperties;
+  private final int mNumMasters;
+  private final int mNumWorkers;
+  private final String mClusterName;
+
+  /** Base directory for storing configuration and logs. */
+  private File mWorkDir;
+  /** Closer for closing all resources that must be closed when the cluster is destroyed. */
+  private Closer mCloser;
+  private List<Master> mMasters;
+  private List<Worker> mWorkers;
+  /** Addresses of all masters. Should have the same size as {@link #mMasters}. */
+  private List<MasterNetAddress> mMasterAddresses;
+  private State mState;
+  private TestingServer mCuratorServer;
+
+  private ExternalCluster(Map<PropertyKey, String> properties, int numMasters, int numWorkers,
+      String clusterName) {
+    mProperties = properties;
+    mNumMasters = numMasters;
+    mNumWorkers = numWorkers;
+    // Add a unique number so that different runs of the same test use different cluster names.
+    mClusterName = clusterName + ThreadLocalRandom.current().nextLong();
```

Comment:
What's the significance of `clusterName`, could it be purely randomly generated?

Reply:
it's useful for identifying which test the cluster is for - in particular when we collect artifacts for failed clusters, this name is used in the tarball name.